mantidproject/mantid
# Add importance sampling according to DiscusMultipleScatteringCorrection

**Description of work.**

Add importance sampling option to DiscusMultipleScatteringCorrection so that it converges more quickly for situations where S(Q) contains spikes eg diffraction

The previous gaussian interpolation that was used to interpolate the S(Q) distribution has been removed because it didn't work very nicely with the importance sampling. It has been replaced by a more simple "flat" interpolation from the previous point in the distribution. This also removes the need to insist on the S(Q) distribution being >0 at all points and having equally spaced points

**To test:**

The following script calculates a multiple scattering correction that uses an S(Q) profile consisting of a single spike at Q=7. The input workspace has detectors spread over a two theta range 0 to 180 degrees, each spectrum has a single bin centred on wavelength=1 angstrom. The wavelength and the spike position corresponds to a two theta angle of 67.7 degrees.

This should give a multiple scattering correction with a peak at two theta=0 and two theta=2*67.7=135.4 degrees corresponding to two scatters with a smooth U shaped curve in between. The detector with two theta=135.4 degrees is attached to spectrum ~680

If you run the script twice with ImportanceSampling set to True and False you should see that the result with "True" is much cleaner and the peaks at 0 and 135.4 degrees are visible with the U shape in between. If you change the peak width to be x10 narrower the result without importance sampling degenerates further.

I suggest you run it on a release build to save a bit of time.

![image](https://user-images.githubusercontent.com/56295817/136770373-4f80bcf8-3fb0-4ffb-a16f-601208bf66f8.png)

The correction is stored in the Scatter_2 workspace that sits inside a workspace group called MSResults.

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from testhelpers import WorkspaceCreationHelper

test_ws = WorkspaceCreationHelper.create2DWorkspaceWithGeographicalDetectors(1, 900, 0.2, 1)
ws_name = "test"
AnalysisDataService.addOrReplace(ws_name, test_ws)

X=[6.9,7.0,7.1]
Y=[0.,1000,0.]
Sofq=CreateWorkspace(DataX=X,DataY=Y,UnitX="MomentumTransfer")

SetSample(InputWorkspace=test_ws,
          Geometry={'Shape': 'FlatPlate', 'Height': 4.0,
                  'Width': 4.0, 'Thick': 1.0,
                  'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': 'Ni'})

DiscusMultipleScatteringCorrection(InputWorkspace=test_ws, SofqWorkspace=Sofq, OutputWorkspace="MSResults",
                                   NeutronPathsSingle=1, NeutronPathsMultiple=10000, NumberScatterings=2,
                                   ImportanceSampling=True)
```

Fixes #32208.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
